### PR TITLE
Registration

### PR DESCRIPTION
**Pull request by andrewseamanco**
- Template for testing
- Filter that works for registration
- Username taken or field not filled functionality
- Username taken endpoint [returns true or false]
- Creates new user object

### CODE REVIEW

File: bait-files/src/main/webapp/WEB-INF/Error404.jsp [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
+    "http://www.w3.org/TR/html4/loose.dtd">
+<html>
+<head>
+<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
+<title>404 Error - Page Not Found</title>
+</head>
+<body>
+    <center>
+        <h1>Sorry, the page you requested were not found.</h1>
```
Comment (sunr1) [inline bait-files/src/main/webapp/WEB-INF/Error404.jsp:12]: very small change but should be "was not found"

File: bait-files/src/main/webapp/register.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+function checkFormInput() {
+    const errorMessageDiv = document.getElementById('error-message-div');
+    removeErrorMessage(errorMessageDiv);
+
+    let username = document.getElementById('username-input').value;
+    fetch('/usernameTaken?username=' + username)
+      .then(response => response.json())
+      .then((usernameTaken) => {
+        if (!allFieldsFilled()) {
+          printError("One or more fields are empty. Please fill out all fields to contine.", errorMessageDiv);
```
Comment (sunr1) [inline bait-files/src/main/webapp/register.js:10]: typo on continue

File: bait-files/src/main/webapp/WEB-INF/Error404.jsp (hunk 1 above)
Comment (sunr1) [inline bait-files/src/main/webapp/WEB-INF/Error404.jsp:12]: very small change but should be "was not found"

File: bait-files/src/main/java/com/google/LoginFilter.java [hunk 3]
```diff
@@ -62,7 +60,7 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
       if (!isRegistered(userService.getCurrentUser().getUserId())
           && !request.getRequestURI().startsWith("/_ah/")) {
         // User is submitting form for registration
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/LoginFilter.java:62]: How does this tell us they're submitting the form for registration? Expand upon this comment.

File: bait-files/src/main/java/com/google/LoginFilter.java [hunk 4]
```diff
@@ -81,7 +79,7 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
     if (request.getRequestURI().endsWith("profile.jsp")
         || request.getRequestURI().endsWith("register.jsp")
         || request.getRequestURI().endsWith("login.jsp")) {
-      System.out.println("here");
+              System.out.println("redirect to profile");
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/LoginFilter.java:82]: Remove :)

File: bait-files/src/main/java/com/google/LoginFilter.java [hunk 5]
```diff
@@ -44,8 +44,6 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
     HttpSession session = request.getSession(false);
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/LoginFilter.java:44]: I can't comment higher up, but this class should have a java doc.

File: bait-files/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 6]
```diff
@@ -23,28 +23,7 @@
   @Override
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/sps/servlets/LoginServlet.java:23]: This class should also have a java doc.

File: bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.List;
+import java.util.Arrays;
+import java.util.Random;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/usernameTaken")
+public class UsernameTakenServlet extends HttpServlet {
```
Comment (kzweller) [inline bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java:30]: Class should have java doc.

File: bait-files/src/main/webapp/WEB-INF/Error404.jsp [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<%@ page language="java" contentType="text/html; charset=ISO-8859-1"
```
Comment (kzweller) [inline bait-files/src/main/webapp/WEB-INF/Error404.jsp:1]: Under what circumstances does this page manifest?

File: bait-files/src/test/java/com/google/sps/UserTest.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
+package com.google.sps;
```
Comment (kzweller) [inline bait-files/src/test/java/com/google/sps/UserTest.java:1]: This file looks out of date? Fix this. Don't have in-progress stuff assigned to us for review :)

File: bait-files/src/main/webapp/WEB-INF/Error404.jsp (hunk 8 above)
Comment (andrewseamanco) [reply bait-files/src/main/webapp/WEB-INF/Error404.jsp]: Removed

File: bait-files/src/main/webapp/WEB-INF/Error404.jsp (hunk 1 above)
Comment (andrewseamanco) [reply bait-files/src/main/webapp/WEB-INF/Error404.jsp]: Removed

File: bait-files/src/main/webapp/register.js (hunk 2 above)
Comment (andrewseamanco) [reply bait-files/src/main/webapp/register.js:10]: Resolved

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 4 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java:82]: Removed

File: bait-files/src/test/java/com/google/sps/UserTest.java (hunk 9 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UserTest.java:1]: Added progress

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 3 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java]: Clarified

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 5 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java:47]: Added javadoc

File: bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java (hunk 7 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java:33]: Added javadoc

File: bait-files/src/main/java/com/google/sps/servlets/LoginServlet.java (hunk 6 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/sps/servlets/LoginServlet.java:34]: Added javadoc

File: bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+   * @param request contains a url parameter with username to check
+   * @param response returns a json response with a boolean value if username is taken
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String username = request.getParameter("username");
+
+    response.setContentType("application/json");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    if (username==null || username.equals("")) {
```
Comment (tambchop) [inline bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java:46]: Formatting: put spaces around double equals.

File: bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        response.getWriter().println(json);
+        return;
+    }
+
+    Query query = new Query(USER_ENTITY)
+      .setFilter(new Query.FilterPredicate(
+        USERNAME_ENTITY_PROPERTY, Query.FilterOperator.EQUAL, username));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    boolean usernameTaken = entity!=null;
```
Comment (tambchop) [inline bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java:57]: Formatting: put spaces around the !=

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 4 above)
Comment (tambchop) [reply bait-files/src/main/java/com/google/LoginFilter.java:85]: So you say "removed"... but I still see the println here... (insert suspicious, chin scratching emoji)

File: bait-files/src/test/java/com/google/sps/UserTest.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  // Database is empty, so username should not be taken
+  @Test
+  public void emptyUsernameTakenTest() throws IOException {
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/UserTest.java:48]: So a note about unit test naming...  There are a lot of conventions, but for the most part what we do on our team is to follow this naming convention:
public void methodBeingTested_setupInfoForThisTest_expectedResult()

So in this first case, I'd go with:
public void doGet_usernameNotInDatabase_returnsNotTaken

Or something like that.  Now when people read the test name, they can see that this is a test of the doGet method, for the case when the proposed username does not already exist in the database, and the expected result is that the response will say that the username is not taken =)

This is a pretty concise and to the point way of writing test names, and it also means that we don't need to write a comment before every test, so you'll be able to delete those.  As in other situations, it's always nice when the names of things in code are clear enough that comments aren't required =)

Also if you're wondering, we do normally avoid using underscores in naming methods and variables (other than constant variables, which are in all uppercase), but the naming of test methods is a specific exception when we very much do use them.

File: bait-files/src/test/java/com/google/sps/UserTest.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+import java.io.*;
+import javax.servlet.http.*;
+import javax.servlet.ServletException;
+import com.google.sps.servlets.UsernameTakenServlet;
+import com.google.sps.servlets.LoginServlet;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class UserTest {
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/UserTest.java:30]: It's a good practice to create one unit test file per class you have and name them with the exact same name as the class you are testing except with the word "Test" added to the end.

So in your case, it looks like here you are testing the UsernameTakenServlet class and the LoginServlet, so I would say that you should create these files instead:
UsernameTakenServletTest.java  (containing class UsernameTakenServletTest)
LoginServlet.java  (containing class LoginServletTest)

File: bait-files/src/test/java/com/google/sps/UserTest.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+        assertEquals(1, ds.prepare(new Query(USER_ENTITY)).countEntities(withLimit(10)));
+  }
+
+    // Add a user and ensure there is a new entry in database
+  @Test
+  public void addUser2() throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    assertTrue(userService.isUserAdmin());
+  }
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/UserTest.java:130]: It looks like all you are testing here is that Google's UserService class works as intended, so I'd say you don't need this test.  I assume Google has already tested this class themselves =)

File: bait-files/src/test/java/com/google/sps/UserTest.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        new UsernameTakenServlet().doGet(request, response);
+
+        verify(request, atLeast(1)).getParameter("username"); // only if you want to verify username was called...
+        writer.flush(); // it may not have been flushed yet...
+        assertTrue(stringWriter.toString().contains("false"));
+  }
+
+  // Username field is null, so return that the username is taken
+  @Test
+  public void nullUsernameTest() throws IOException {
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/UserTest.java:67]: Same comment about test naming.. maybe doGet_nullRequestedUsername_returnsTaken.

As a final note, the word "returns" in the last part is kind of optional since with this three part naming style, it can be kind of understood that the last part means what is returned.

File: bait-files/src/test/java/com/google/sps/UserTest.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+        PrintWriter writer = new PrintWriter(stringWriter);
+
+        when(response.getWriter()).thenReturn(writer);
+        new UsernameTakenServlet().doGet(request, response);
+
+        verify(request, atLeast(1)).getParameter("username"); // only if you want to verify username was called...
+        writer.flush(); // it may not have been flushed yet...
+        assertTrue(stringWriter.toString().contains("true"));
+  }
+
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/UserTest.java:103]: You've done a great job at thinking to test the different edge cases in the UsernameTakenServlet!

I'd just add one final test for when the user actually does exist in the database:
doGet_usernameExistsAlready_returnsTaken

File: bait-files/src/test/java/com/google/sps/UserTest.java [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+  // Add a user and ensure there is a new entry in database
+  @Test
+  public void addUser() throws IOException, ServletException {
+        HttpServletRequest request = mock(HttpServletRequest.class);       
+        HttpServletResponse response = mock(HttpServletResponse.class);
+
+        UserService userService = UserServiceFactory.getUserService();
+
+        System.out.println(userService.getCurrentUser());
+        System.out.println("==================");
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/UserTest.java:113]: Remove these =D

File: bait-files/src/test/java/com/google/sps/UserTest.java [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+        new UsernameTakenServlet().doGet(request, response);
+
+        verify(request, atLeast(1)).getParameter("username"); // only if you want to verify username was called...
+        writer.flush(); // it may not have been flushed yet...
+        assertTrue(stringWriter.toString().contains("true"));
+  }
+
+  // Add a user and ensure there is a new entry in database
+  @Test
+  public void addUser() throws IOException, ServletException {
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/UserTest.java:106]: As mentioned way above, I'd move this into a new test file called LoginServletTest.java, and naming wise you could name this test something like:
doPost_addsUserToDatabase

I'd say it's fine to break the three part naming convention when you only need one test and therefore don't have a variety of setup conditions.  If you did want a three part name, you could always go with something like:
doPost_forAnyInput_addsUserToDatabase

File: bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java (hunk 10 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java:46]: Added spacing

File: bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java (hunk 11 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/sps/servlets/UsernameTakenServlet.java:57]: Added spacing

File: bait-files/src/test/java/com/google/sps/UserTest.java (hunk 12 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UserTest.java]: Changed name

File: bait-files/src/test/java/com/google/sps/UserTest.java (hunk 14 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UserTest.java]: Removed

File: bait-files/src/test/java/com/google/sps/UserTest.java (hunk 15 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UserTest.java]: renamed

File: bait-files/src/test/java/com/google/sps/UserTest.java (hunk 17 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UserTest.java]: Removed

File: bait-files/src/test/java/com/google/sps/UserTest.java (hunk 18 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UserTest.java]: Renamed

File: bait-files/src/test/java/com/google/sps/UserTest.java (hunk 13 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UserTest.java:30]: Separated concerns :)

File: bait-files/src/test/java/com/google/sps/UserTest.java (hunk 16 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UserTest.java]: Added a new test

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 4 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java:85]: It says it's removed on my local cloud terminal :/

File: bait-files/src/test/java/com/google/sps/UsernameTakenTest.java [hunk 19]
```diff
@@ -27,11 +27,11 @@
 import org.junit.runners.JUnit4;
 
 @RunWith(JUnit4.class)
-public final class UserTest {
+public final class UsernameTakenTest {
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/UsernameTakenTest.java:30]: Nit: Include the word "Servlet" in this class and file name: UsernameTakenServletTest.  To make it super obvious that the class you are testing is UsernameTakenServlet.

File: bait-files/src/test/java/com/google/sps/LoginTest.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+import java.io.*;
+import javax.servlet.http.*;
+import javax.servlet.ServletException;
+import com.google.sps.servlets.UsernameTakenServlet;
+import com.google.sps.servlets.LoginServlet;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class LoginTest {
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/LoginTest.java:28]: Nit: Include the word "Servlet" in this class and file name: LoginServletTest.  To make it super obvious that the class you are testing is LoginServlet.

File: bait-files/src/test/java/com/google/sps/UsernameTakenTest.java (hunk 19 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/UsernameTakenTest.java]: Rename to servlet

File: bait-files/src/test/java/com/google/sps/LoginTest.java (hunk 20 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/LoginTest.java]: Rename to servlet

File: bait-files/src/main/java/com/google/LoginFilter.java (hunk 4 above)
Comment (andrewseamanco) [reply bait-files/src/main/java/com/google/LoginFilter.java]: Okay it's gone now

File: bait-files/src/test/java/com/google/sps/LoginTest.java [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  // Add a user and ensure there is a new entry in database
+  @Test
+  public void doPost_addsUserSuccessfully_returnsUserInDatabase() throws IOException, ServletException {
+        HttpServletRequest request = mock(HttpServletRequest.class);
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/LoginTest.java]: Naming nit: the second part of the test name is meant to describe the setup or input of the particular test case.  "addsUserSuccessfully" sounds like a result... i.e. something that should be in the last part.

Maybe something like:
doPost_forNewUser_addsUserToDatabase

Or like I said before, if it feels overly obvious to include the middle part (the input / scenario), you could just have the first and last parts:
doPost_addsUserToDatabase

Although having said all that, when I look at the content of the actual doPost method, I see that the last thing it does is to apply a redirect on the response, so it could actually be good to test that.  You do have a mock of the response object in this test.  I forget the syntax for this, but from looking at Mockito documentation (https://github.com/mockito/mockito/wiki#examples) it could be something like:
verify(response).sendRedirect("/profile.jsp");

Then your test could be called something like:
doPost_addsUserAndRedirects or
doPost_forNewUser_addsUserAndRedirects

File: bait-files/WEB-INF/appengine-generated/datastore-indexes-auto.xml [hunk 22]
```diff
@@ -1,4 +1,4 @@
-<!-- Indices written at Tue, 14 Jul 2020 19:16:49 UTC -->
+<!-- Indices written at Tue, 14 Jul 2020 22:39:27 UTC -->
 
 <datastore-indexes/>
```
Comment (tambchop) [inline bait-files/WEB-INF/appengine-generated/datastore-indexes-auto.xml:4]: This file as well as local_db.bin don't seem to be necessary to the GitHub repository.  It looks like they are autogenerated any time that anybody runs the code on their own computer.  I think it should be safe to set your repository to never commit anything in the appengine-generated folder.

To do that, you just need to modify the file called .gitignore in the root directory.  I think adding the following line to the file should work:
bait-files/WEB-INF/appengine-generated/

If it does work, you should see that these two files should no longer be in the commit.

File: bait-files/src/test/java/com/google/sps/LoginTest.java (hunk 21 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/LoginTest.java]: Renamed tests

File: bait-files/WEB-INF/appengine-generated/datastore-indexes-auto.xml (hunk 22 above)
Comment (andrewseamanco) [reply bait-files/WEB-INF/appengine-generated/datastore-indexes-auto.xml:4]: Removed

File: bait-files/src/test/java/com/google/sps/LoginServletTest.java [hunk 23]
```diff
@@ -55,5 +55,6 @@ public void doPost_addsUserSuccessfully_returnsUserInDatabase() throws IOExcepti
 
         DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
         assertEquals(1, ds.prepare(new Query(USER_ENTITY)).countEntities(withLimit(10)));
+        verify(response, atLeast(1)).sendRedirect("/profile.jsp"); // only if you want to verify username was called...
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/LoginServletTest.java:58]: I don't understand the comment here.  The line of code seems like it speaks for itself: verify that a redirect to /profile.jsp happens.  I don't think a comment is necessary =)

File: bait-files/src/test/java/com/google/sps/LoginServletTest.java [hunk 24]
```diff
@@ -43,7 +43,7 @@ public void tearDown() {
 
   // Add a user and ensure there is a new entry in database
   @Test
-  public void doPost_addsUserSuccessfully_returnsUserInDatabase() throws IOException, ServletException {
+  public void doPost_forNewUser_returnsUserInDatabase() throws IOException, ServletException {
```
Comment (tambchop) [inline bait-files/src/test/java/com/google/sps/LoginServletTest.java:46]: One more naming nitpick =P
...Looking at the final version of this test, I see that we are asserting/verifying two things:
1)  The user now exists in the database.
2)  A redirect happens.

So I'd change the last part of this test name to reflect that: _createsUserAndRedirects.
So: doPost_forNewUser_createsUserAndRedirects

File: bait-files/src/test/java/com/google/sps/LoginServletTest.java (hunk 23 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/LoginServletTest.java]: removed comment

File: bait-files/src/test/java/com/google/sps/LoginServletTest.java (hunk 24 above)
Comment (andrewseamanco) [reply bait-files/src/test/java/com/google/sps/LoginServletTest.java]: renamed
